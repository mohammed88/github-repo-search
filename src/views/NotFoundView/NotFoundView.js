import React, {Fragment} from 'react';
import styled from 'styled-components';

import Header from '../../components/Header/';
import Footer from '../../components/Footer/';
import {H1, Text18} from '../../ui/styles/typography';
import Link from '../../ui/Link';
import Spacing from '../../ui/styles/spacing';
import GitHubText from '../../ui/icons/github-text';
import {darkOrange} from '../../ui/styles/colors';

const Body = styled.div`
	box-sizing: border-box;
	height: 800px;
	padding: 200px;
	line-height: 1.8;
`;

const Message = styled.div`
	text-align: center;
	width: 400px;
	margin: auto;
	${Text18};
`;

const HomeLink = styled(Link)`
	${Text18};
`;

const NotFoundHeader = styled(H1)`
	margin: ${Spacing.large} 0;
`

const GitHubLogo = styled(GitHubText)`
	fill: ${darkOrange};
`;

function NotFoundView() {
	return (
		<Fragment>
			<Header />
			<Body data-component-type="PageNotFoundMessage">
				<Message>
					<GitHubLogo width={200} height={75}/>
					<NotFoundHeader>404 - Not Found</NotFoundHeader>
					Sorry, the page you are looking for does not exist.
					You can always go back to the <HomeLink href="/">homepage.</HomeLink>
				</Message>
			</Body>
			<Footer />
		</Fragment>
	);
}

export default NotFoundView;