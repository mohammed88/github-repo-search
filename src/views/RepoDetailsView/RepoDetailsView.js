import React from 'react';
import styled from 'styled-components';

import Header from '../../components/Header/';
import Footer from '../../components/Footer/';
import Spacing from '../../ui/styles/spacing';
import {H2, Text12, Text14} from '../../ui/styles/typography';
import NotFoundView from '../../views/NotFoundView';
import BASE_URL from '../constants';
import GitBranch from '../../ui/icons/git-branch';
import Repo from '../../ui/icons/repo';
import InfoIcon from '../../components/InfoIcon/';
import Link from '../../ui/Link/';

const Wrapper = styled.div`
	display: flex;
	flex-direction: column;
	height: 100vh;
`;

const Body = styled.div`
	flex: 1 1 0;
	padding: ${Spacing.large};
	overflow: scroll;
	margin: 0 auto;
	box-sizing: border-box;
	width: 800px;

	@media (max-width: 800px) {
		width: 100%;
	}
`;

const BranchesList = styled.ul`
	display: flex;
	flex-direction: column;
	flex-wrap: wrap;
	list-style-type: none;
	line-height: 1.5;
	padding: 0;
	height: 90%;

	@media (max-width: 760px) {
		flex-wrap: nowrap;
	}
`;

const Title = styled.div`
	display: flex;
	align-items: center;
`;

const SubTitle = styled.div`
	${Text14};
	margin: 0 ${Spacing.medium};
	font-style: italic;

`;

const RepoIcon = styled(Repo)`
	margin-right: ${Spacing.small};
`;

const BranchLink = styled(Link)`
	${Text12};
	font-weight: bold;
`;

class RepoDetailsView extends React.Component {
	state = {
		branches: [],
		repoNotFound: false,
	};

	async fetchData(url) {
		const response = await fetch(url);
		return await response.json();
	}

	async fetchRepoData() {
		try {
			const {owner, repo} = this.props.match.params;
			const url = `${BASE_URL}/repos/${owner}/${repo}/branches`;

			const branches = await this.fetchData(url);

			if (branches.message) {
				throw(branches.message);
			}

			this.setState({branches, repoNotFound: false});
		} catch(e) {
			this.setState({repoNotFound: true});
		}
	}

	componentDidMount() {
		this.fetchRepoData();
	}

	render() {
		const {branches, repoNotFound} = this.state;
		const {owner, repo} = this.props.match.params;

		if (repoNotFound) {
		
			return <NotFoundView />;
		}

		return(
			<Wrapper>
				<Header />
				<Body>
					<Title>
						<RepoIcon />
						<H2>{owner} / {repo} </H2>
						<SubTitle>({branches.length} branches)</SubTitle>
					</Title>
					<BranchesList>
						{branches.length === 0 && <div>Loading...</div>}
						{branches.map(({name, commit}, index) => 
							<li key={index} data-component-type="BranchListItem">
								<InfoIcon
									icon={<GitBranch />} 
									text={
										<BranchInfo 
											owner={owner} 
											repo={repo} 
											commit={commit} 
											name={name} 
										/>
									}
								/>
							</li>
						)}
					</BranchesList>
				</Body>
				<Footer />
			</Wrapper>
		);
	}
}

function BranchInfo({owner, repo, commit, name}) {
	return (
		<BranchLink
			href={`http://github.com/${owner}/${repo}/commit/${commit.sha}`}
			target="_blank">
			{name}
		</BranchLink>
	);
}

export default RepoDetailsView;