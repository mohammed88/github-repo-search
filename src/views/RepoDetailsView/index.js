import RepoDetailsView from './RepoDetailsView';

export default RepoDetailsView;