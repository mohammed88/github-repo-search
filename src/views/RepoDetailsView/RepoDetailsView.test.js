import React from 'react';
import {shallow, mount} from 'enzyme';
import { StaticRouter } from 'react-router-dom';

import RepoDetailsView from './RepoDetailsView';
import NotFoundView from '../../views/NotFoundView';

describe('RepoDetailsView', () => {
	const match = {params: {owner: '', repo: ''}};

	it('should render', () => {
		const wrapper = shallow(
			<RepoDetailsView match={match} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should render details when data is fetched successfully', (done) => {
		const mockData = [{}, {}];
		RepoDetailsView.prototype.fetchData = jest.fn().mockReturnValue(Promise.resolve(mockData));

		const wrapper = shallow(<RepoDetailsView match={match} />);

		process.nextTick(() => {
			expect(wrapper.state().branches.length).toEqual(mockData.length);
			expect(wrapper.find(NotFoundView).length).toBe(0);
			done();
		});
	});

	it('should render Not Found message when data fetching fails', (done) => {
		RepoDetailsView.prototype.fetchData = jest.fn().mockReturnValue(Promise.reject());
		
		const wrapper = shallow(<RepoDetailsView match={match} />);

		process.nextTick(() => {
			expect(wrapper.find(NotFoundView).length).toBe(1);
			done();
		});
	});
});