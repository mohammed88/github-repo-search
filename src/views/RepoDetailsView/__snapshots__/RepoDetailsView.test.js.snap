// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoDetailsView should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RepoDetailsView
    match={
      Object {
        "params": Object {
          "owner": "",
          "repo": "",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header />,
        <ForwardRef>
          <ForwardRef>
            <ForwardRef />
            <ForwardRef>
              
               / 
              
               
            </ForwardRef>
            <ForwardRef>
              (
              0
               branches)
            </ForwardRef>
          </ForwardRef>
          <ForwardRef>
            <div>
              Loading...
            </div>
          </ForwardRef>
        </ForwardRef>,
        <Footer />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef>
              <ForwardRef />
              <ForwardRef>
                
                 / 
                
                 
              </ForwardRef>
              <ForwardRef>
                (
                0
                 branches)
              </ForwardRef>
            </ForwardRef>,
            <ForwardRef>
              <div>
                Loading...
              </div>
            </ForwardRef>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef />,
                <ForwardRef>
                  
                   / 
                  
                   
                </ForwardRef>,
                <ForwardRef>
                  (
                  0
                   branches)
                </ForwardRef>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-chPdSV",
                    "isStatic": true,
                    "rules": Array [
                      "
	margin-right: ",
                      "8px",
                      ";
",
                    ],
                  },
                  "displayName": "Styled(RepoIcon)",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-chPdSV",
                  "target": [Function],
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                  Symbol(Symbol.hasInstance): [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    "",
                    " / ",
                    "",
                    " ",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "",
                  " / ",
                  "",
                  " ",
                ],
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-ifAKCX",
                    "isStatic": true,
                    "rules": Array [
                      "
	font-family: ",
                      "'Roboto', sans-serif",
                      ";
	font-size: 18px;
	font-weight: bold;
	margin: 0;
",
                    ],
                  },
                  "displayName": "styled.h2",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-ifAKCX",
                  "target": "h2",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    "(",
                    0,
                    " branches)",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "(",
                  0,
                  " branches)",
                ],
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-kAzzGY",
                    "isStatic": true,
                    "rules": Array [
                      "
	",
                      "
	font-family: ",
                      "'Roboto', sans-serif",
                      ";
	font-size: 14px;
",
                      ";
	margin: 0 ",
                      "12px",
                      ";
	font-style: italic;

",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-kAzzGY",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-cSHVUG",
                "isStatic": true,
                "rules": Array [
                  "
	display: flex;
	align-items: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-cSHVUG",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <div>
                  Loading...
                </div>,
                Array [],
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Loading...",
                },
                "ref": null,
                "rendered": "Loading...",
                "type": "div",
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-jzJRlG",
                "isStatic": true,
                "rules": Array [
                  "
	display: flex;
	flex-direction: column;
	flex-wrap: wrap;
	list-style-type: none;
	line-height: 1.5;
	padding: 0;
	height: 90%;

	@media (max-width: 760px) {
		flex-wrap: nowrap;
	}
",
                ],
              },
              "displayName": "styled.ul",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-jzJRlG",
              "target": "ul",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-fjdhpX",
            "isStatic": true,
            "rules": Array [
              "
	flex: 1 1 0;
	padding: ",
              "24px",
              ";
	overflow: scroll;
	margin: 0 auto;
	box-sizing: border-box;
	width: 800px;

	@media (max-width: 800px) {
		width: 100%;
	}
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-fjdhpX",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-jTzLTM",
        "isStatic": true,
        "rules": Array [
          "
	display: flex;
	flex-direction: column;
	height: 100vh;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-jTzLTM",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header />,
          <ForwardRef>
            <ForwardRef>
              <ForwardRef />
              <ForwardRef>
                
                 / 
                
                 
              </ForwardRef>
              <ForwardRef>
                (
                0
                 branches)
              </ForwardRef>
            </ForwardRef>
            <ForwardRef>
              <div>
                Loading...
              </div>
            </ForwardRef>
          </ForwardRef>,
          <Footer />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef>
                <ForwardRef />
                <ForwardRef>
                  
                   / 
                  
                   
                </ForwardRef>
                <ForwardRef>
                  (
                  0
                   branches)
                </ForwardRef>
              </ForwardRef>,
              <ForwardRef>
                <div>
                  Loading...
                </div>
              </ForwardRef>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <ForwardRef />,
                  <ForwardRef>
                    
                     / 
                    
                     
                  </ForwardRef>,
                  <ForwardRef>
                    (
                    0
                     branches)
                  </ForwardRef>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-chPdSV",
                      "isStatic": true,
                      "rules": Array [
                        "
	margin-right: ",
                        "8px",
                        ";
",
                      ],
                    },
                    "displayName": "Styled(RepoIcon)",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-chPdSV",
                    "target": [Function],
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                    Symbol(Symbol.hasInstance): [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      "",
                      " / ",
                      "",
                      " ",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "",
                    " / ",
                    "",
                    " ",
                  ],
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-ifAKCX",
                      "isStatic": true,
                      "rules": Array [
                        "
	font-family: ",
                        "'Roboto', sans-serif",
                        ";
	font-size: 18px;
	font-weight: bold;
	margin: 0;
",
                      ],
                    },
                    "displayName": "styled.h2",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-ifAKCX",
                    "target": "h2",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      "(",
                      0,
                      " branches)",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "(",
                    0,
                    " branches)",
                  ],
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-kAzzGY",
                      "isStatic": true,
                      "rules": Array [
                        "
	",
                        "
	font-family: ",
                        "'Roboto', sans-serif",
                        ";
	font-size: 14px;
",
                        ";
	margin: 0 ",
                        "12px",
                        ";
	font-style: italic;

",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-kAzzGY",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  },
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-cSHVUG",
                  "isStatic": true,
                  "rules": Array [
                    "
	display: flex;
	align-items: center;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-cSHVUG",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <div>
                    Loading...
                  </div>,
                  Array [],
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Loading...",
                  },
                  "ref": null,
                  "rendered": "Loading...",
                  "type": "div",
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-jzJRlG",
                  "isStatic": true,
                  "rules": Array [
                    "
	display: flex;
	flex-direction: column;
	flex-wrap: wrap;
	list-style-type: none;
	line-height: 1.5;
	padding: 0;
	height: 90%;

	@media (max-width: 760px) {
		flex-wrap: nowrap;
	}
",
                  ],
                },
                "displayName": "styled.ul",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-jzJRlG",
                "target": "ul",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-fjdhpX",
              "isStatic": true,
              "rules": Array [
                "
	flex: 1 1 0;
	padding: ",
                "24px",
                ";
	overflow: scroll;
	margin: 0 auto;
	box-sizing: border-box;
	width: 800px;

	@media (max-width: 800px) {
		width: 100%;
	}
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-fjdhpX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-jTzLTM",
          "isStatic": true,
          "rules": Array [
            "
	display: flex;
	flex-direction: column;
	height: 100vh;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-jTzLTM",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
